dcore/advection.py

```diff
@@ -202,3 +202,94 @@ def apply(self, x_in, x_out):
         self.xdg_in.interpolate(x_in)
         super(EmbeddedDGAdvection, self).apply(self.xdg_in, self.xdg_out)
         x_out.project(self.xdg_out)
+
+
+class EulerPoincareForm(Advection):
+
+    def __init__(self, state, V):
+        super(EulerPoincareForm, self).__init__(state)
+
+        dt = state.timestepping.dt
+        w = TestFunction(V)
+        u = TrialFunction(V)
+        self.u0 = Function(V)
+        ustar = 0.5*(self.u0 + u)
+        n = FacetNormal(state.mesh)
+        Upwind = 0.5*(sign(dot(self.ubar, n))+1)
+
+        # define surface measure and terms involving perp differently
+        # for slice (i.e. if V.extruded is True) and shallow water
+        # (V.extruded is False)
+        if V.extruded:
+            surface_measure = (dS_h + dS_v)
+            perp = lambda u: as_vector([-u[1], u[0]])
+            perp_u_upwind = Upwind('+')*perp(ustar('+')) + Upwind('-')*perp(ustar('-'))
+        else:
+            surface_measure = dS
+            outward_normals = CellNormal(state.mesh)
+            perp = lambda u: cross(outward_normals, u)
+            perp_u_upwind = Upwind('+')*cross(outward_normals('+'),ustar('+')) + Upwind('-')*cross(outward_normals('-'),ustar('-'))
+
+        Eqn = (
+            (inner(w, u-self.u0)
+             - dt*inner(w, div(perp(ustar))*perp(self.ubar))
+             - dt*div(w)*inner(ustar, self.ubar))*dx
+            - dt*inner(jump(inner(w, perp(self.ubar)), n), perp_u_upwind)*surface_measure
+            + dt*jump(inner(w, perp(self.ubar))*perp(ustar), n)*surface_measure
+        )
+        a = lhs(Eqn)
+        L = rhs(Eqn)
+        self.u1 = Function(V)
+        uproblem = LinearVariationalProblem(a, L, self.u1)
+        self.usolver = LinearVariationalSolver(uproblem)
+
+    def apply(self, x_in, x_out):
+        self.u0.assign(x_in)
+        self.usolver.solve()
+        x_out.assign(self.u1)
+
+
+class SUPGAdvection(Advection):
+
+    def __init__(self, state, V, direction=None, supg_params=None):
```

Comment:
'direction' is supposed to specify the direction in which we're not applying SUPG - so the entry for tau is set to zero for that direction. I've just realised I also need to change the surface measure, depending on 'direction'.